### PR TITLE
[Intrinsics][PreISelInstrinsicLowering] llvm.memcpy.inline length no longer needs to be constant

### PR DESCRIPTION
**Pull request by asb**
Following on from the discussion in
https://discourse.llvm.org/t/rfc-introducing-an-llvm-memset-pattern-inline-intrinsic/79496 and the equivalent change for llvm.memset.inline (#95397), this removes the requirement that the length of llvm.memcpy.inline is constant. PreISelInstrinsicLowering will expand llvm.memcpy.inline with non-constant lengths, while the codegen path for constant lengths is left unaltered.

### CODE REVIEW

File: llvm/test/Transforms/PreISelIntrinsicLowering/X86/memcpy-inline-non-constant-len.ll [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+define void @memcpy_32(ptr %dst, ptr %src) nounwind {
+; CHECK-LABEL: define void @memcpy_32(
+; CHECK-SAME: ptr [[DST:%.*]], ptr [[SRC:%.*]]) #[[ATTR0:[0-9]+]] {
+; CHECK-NEXT:    tail call void @llvm.memcpy.inline.p0.p0.i64(ptr [[DST]], ptr [[SRC]], i64 32, i1 false)
+; CHECK-NEXT:    ret void
+;
+  tail call void @llvm.memcpy.inline.p0.p0.i64(ptr %dst, ptr %src, i64 32, i1 0)
+  ret void
+}
+
```
Comment (arsenm) [inline llvm/test/Transforms/PreISelIntrinsicLowering/X86/memcpy-inline-non-constant-len.ll:14]: Test a volatile case to make sure that doesn't get dropped?

Comment (asb) [reply llvm/test/Transforms/PreISelIntrinsicLowering/X86/memcpy-inline-non-constant-len.ll:16]: Done.

Comment (arsenm) [reply llvm/test/Transforms/PreISelIntrinsicLowering/X86/memcpy-inline-non-constant-len.ll:16]: I meant more in the case that gets expanded, this was recently broken for memmove and fixed in d37e7ec2c5c3ecfd6a4612c1798d62d343ceb254

Comment (asb) [reply llvm/test/Transforms/PreISelIntrinsicLowering/X86/memcpy-inline-non-constant-len.ll:16]: That makes more sense - I've added that now to the memcpy_x case below.

**Review by arsenm — APPROVED** [review]

**Pull request merged**